core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+                                '</div>';
+      }
+      CKEDITOR.plugins.add(ckName, {
+        init: function(editor) {
+          // Create the widget itself.
+          editor.widgets.add(ckName, {
+            button: componentDefn.tooltip,
+            inline: isInline,
+            template: componentTemplate,
+            edit: function(event) {
+              editor.fire('lockSnapshot', {
+                dontUpdate: true
+              });
+              // Prevent default action since we are using our own edit modal.
+              event.cancel();
+              // Save this for creating the widget later.
+              var container = this.wrapper.getParent(true);
+              var that = this;
+              var customizationArgs = {};
+              customizationArgSpecs.forEach(function(spec) {
+                customizationArgs[spec.name] = that.data[spec.name] ||
+                                               spec.default_value;
+              });
+
+              rteHelperService._openCustomizationModal(
+                customizationArgSpecs,
+                customizationArgs,
+                function(customizationArgsDict) {
+                  for (var arg in customizationArgsDict) {
+                    if (customizationArgsDict.hasOwnProperty(arg)) {
+                      that.setData(arg, customizationArgsDict[arg]);
+                    }
+                  }
+
+                  if (!that.isReady()) {
+                    // Actually create the widget, if we have not already.
+                    editor.widgets.finalizeCreation(container);
+                  }
+                  // Need to manually $compile so the directive renders.
+                  $compile($(that.element.$).contents())($rootScope);
+                  $timeout(function() {
+                    if (isInline) {
+                      // Move caret after the newly created widget.
+                      var range = editor.createRange();
+                      var widgetContainer = that.element.getParent();
+                      range.moveToPosition(
+                        widgetContainer, CKEDITOR.POSITION_AFTER_END);
+                      editor.getSelection().selectRanges([range]);
+                      $timeout(function() {
+                        editor.fire('unlockSnapshot');
+                        editor.fire('saveSnapshot');
+                      });
+                    } else {
+                      editor.fire('unlockSnapshot');
+                      editor.fire('saveSnapshot');
+                    }
+                  });
+                },
+                function() {},
+                function() {});
+            },
+            downcast: function(element) {
+              element.children[0].setHtml('');
+              return element.children[0];
+            },
+            upcast: function(element) {
+              // This is how the widget is recognized by CKEditor.
```

Comment:
Maybe give more detail about downcast/upcast -- what do their return values signify? I tried searching for "CKEditor downcast" and couldn't easily find documentation for it that didn't explain downcasting in terms of downcasting... (!)